ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/query/BlockchainQueries.java

```diff
@@ -483,100 +487,34 @@ private TransactionWithMetadata transactionByHeaderAndIndex(
    */
   public List<LogWithMetadata> matchingLogs(
       final long fromBlockNumber, final long toBlockNumber, final LogsQuery query) {
-    if (fromBlockNumber > toBlockNumber || toBlockNumber > headBlockNumber()) {
-      return Lists.newArrayList();
-    }
-    List<LogWithMetadata> matchingLogs = Lists.newArrayList();
-    for (long blockNumber = fromBlockNumber; blockNumber <= toBlockNumber; blockNumber++) {
-      final Hash blockhash = blockchain.getBlockHashByNumber(blockNumber).get();
-      final boolean logHasBeenRemoved = !blockchain.blockIsOnCanonicalChain(blockhash);
-      final List<TransactionReceipt> receipts = blockchain.getTxReceipts(blockhash).get();
-      final List<Transaction> transaction =
-          blockchain.getBlockBody(blockhash).get().getTransactions();
-      matchingLogs =
-          generateLogWithMetadata(
-              receipts,
-              blockNumber,
-              query,
-              blockhash,
-              matchingLogs,
-              transaction,
-              logHasBeenRemoved);
+    try {
+      return LongStream.rangeClosed(fromBlockNumber, toBlockNumber)
+          .mapToObj(i -> matchingLogs(blockchain.getBlockHashByNumber(i).get(), query))
+          .flatMap(Collection::stream)
+          .collect(Collectors.toList());
+    } catch (NoSuchElementException nsee) {
+      return Collections.emptyList();
```

Comment:
I don't think we want to return an empty list on `NoSuchElementException`.  If `toBlockNumber` exceeds the length of the chain we'd expect this exception.  What about just a simple for loop over the blocks, with a break if no block at the current number is found?  Also, can we add some test coverage for this case?

Reply:
I don't understand what the break is for. If we expect the `NoSuchElementException` then couldn't we just drop the try and catch and let it throw when `.get()` is called?

Reply:
Suggesting something like:
```
public List<LogWithMetadata> matchingLogs(
      final long fromBlockNumber, final long toBlockNumber, final LogsQuery query) {
    if (fromBlockNumber > toBlockNumber || toBlockNumber > headBlockNumber()) {
      return Lists.newArrayList();
    }
    List<LogWithMetadata> matchingLogs = Lists.newArrayList();
    for (long blockNumber = fromBlockNumber; blockNumber <= toBlockNumber; blockNumber++) {
      final Optional<Hash> maybeBlockHash = blockchain.getBlockHashByNumber(blockNumber);
      if (maybeBlockHash.isEmpty()) {
        break;
      }

      final Hash blockHash = maybeBlockHash.get();
      final boolean logHasBeenRemoved = !blockchain.blockIsOnCanonicalChain(blockHash);
      final List<TransactionReceipt> receipts = blockchain.getTxReceipts(blockHash).get();
      final List<Transaction> transaction =
          blockchain.getBlockBody(blockHash).get().getTransactions();
      matchingLogs =
          generateLogWithMetadata(
              receipts,
              blockNumber,
              query,
              blockHash,
              matchingLogs,
              transaction,
              logHasBeenRemoved);
    }
    return matchingLogs;
  }
```

It's not really an error is someone queries for blocks 10-20, and our chain head is at 15.  In that case, we should return the results for blocks 10 - 15.

Reply:
You wanted to merge the methods as well? Did you make these changes on master? What do you think about this minimal change?

```
  public List<LogWithMetadata> matchingLogs(
      final long fromBlockNumber, final long toBlockNumber, final LogsQuery query) {
    return LongStream.rangeClosed(fromBlockNumber, toBlockNumber)
        .mapToObj(blockchain::getBlockHashByNumber)
        .takeWhile(Optional::isPresent)
        .flatMap(definitelyHash -> matchingLogs(definitelyHash.get(), query).stream())
        .collect(Collectors.toList());
  }
```

Reply:
I did not push any changes - your suggestion LGTM :D